spec/iot-js-spec.md

```diff
@@ -4,331 +4,476 @@ IoT Web API
+- If the functionality is not supported, reject ```promise``` with ```NotSupportedError```.
+- If there is no permission to use the method, reject ```promise``` with ```SecurityError```.
+- Send a direct discovery request ```GET /oic/p``` with the given id (which can be either a device UUID or a device URL, and wait for the answer.
+- If there is an error during the request, reject ```promise``` with that error.
+- When the answer is received, resolve ```promise``` with an ```OicPlatform``` object created from the response.
+
+### OIC Presence
+The client API for accessing OIC Presence functionality.
+```javascript
+[NoInterfaceObject]
+interface OicPresence: EventTarget {
+  Promise<void> subscribe(optional USVString url);
+  Promise<void> unsubscribe(optional USVString url);
+  attribute EventHandler<OicDeviceChangeEvent> ondevicechange;
 };
 
-typedef   enum { “init”, “observe”, “create”, “retrieve”, “update”, “delete” } OicMethod;
-
-interface OicRequestEvent : Event {
-  readonly attribute OicMethod type;
-  readonly attribute unsigned long requestId;
-  readonly attribute USVString source;  // device uuid of the OIC client making the request
-  readonly attribute USVString target;  // id of the target resource
-  readonly attribute OicResourceRepresentation? res;  // for “create” and “update” only
-  readonly attribute sequence<DOMString> updatedPropertyNames;  // for “update” only
-  readonly attribute HeaderOption[] headerOptions;  // a hook for flags
-  readonly attribute QueryOption[] queryOptions;  // a hook for extra functionality
-  Promise<void> sendResponse(optional OicResource? resource);  // for create, observe, retrieve
-      // reuses request info (type, requestId, source, target) to construct response,
-      // sends back “ok”, plus the resource object if applicable
-  Promise<void> sendError(Error error);
-      // reuses request info (type, requestId, source, target) to construct response,
-      // sends an Error, where error.message maps to OIC errors, see later
+// event received for device presence notifications
+interface OicDeviceChangeEvent : Event {
+  readonly attribute enum { "added", "deleted", "changed" } type;
+  readonly attribute OicDevice device;
 };
+```
+When the ```subscribe()``` method is invoked, turn on presence listening for the specified ```deviceId```, or if it is not specified, for any device. The Promise may be rejected with ```NotSupportedError```.
 
-interface OicResourceChangedEvent : Event {
-  enum { “add”, “update”, “remove” } type;
-  OicResource resource;
-  DOMString[] updatedPropertyNames; // only for updates, for the rest empty
+### OIC Client
+The OIC Client API provides functionality to access (CRUD) remote resources.
+```javascript
+[NoInterfaceObject]
+interface OicClient {
+  Promise<OicResource> create(OicResourceInit resource);
+  Promise<void> retrieve(OicResourceId id, optional Dictionary parameters);
+  Promise<void> update(OicResourceInit resource);  // partial dictionary
```

Comment:
We need to check that - it would not be very efficient to send the whole resource all again even if one property was changed. The OIC Core spec somehow is not very clear on that.